src/gutter-container.coffee

```diff
@@ -0,0 +1,131 @@
+{Model} = require 'theorist'
+{CompositeDisposable, Emitter} = require 'event-kit'
+{Gutter} = require './gutter'
+
+# This class encapsulates the logic for adding and modifying the gutters on a
+# TextEditor.
+
+module.exports =
+class GutterContainer extends Model
+  # @param textEditor (required) The TextEditor that this GutterContainer is
+  #   associated with.
+  constructor: (textEditor) ->
+    @textEditor = textEditor
+    @leftGutters = []
+    @rightGutters = []
+    @defaultGutter = null
+    @emitter = new Emitter
+
+  destroy: ->
+    @leftGutters = null
+    @rightGutters = null
+    @defaultGutter = null
+    @emitter.dispose()
+    super()
+
+  # Adds a gutter to the left side of the line numbers. Returns a Gutter instance.
+  # @param gutterName (required) a unique string that identifies this gutter in
+  #   the current TextEditor. If the name is already in use, an error is thrown.
+  # @param options (optional) An object with the following fields:
+  #   priority: Determines stacking order between gutters. Lower priority items
+  #     are forced closer to the edges of the window. (default: 100)
+  #   visibleInitially: boolean specifying whether the gutter is visible
+  #     after being created. Defaults to true.
+  addLeftGutter: (gutterName, options) ->
+    options = options ? {}
+    options.location = 'left'
+    @_addSideGutter gutterName, options, @leftGutters
+
+  # Adds a gutter to the right side of the line numbers. Returns a Gutter instance.
+  # @param are the same as for 'addLeftGutter'
+  addRightGutter: (gutterName, options) ->
+    options = options ? {}
+    options.location = 'right'
+    @_addSideGutter gutterName, options, @rightGutters
+
+  _createGutter: (gutterName, options) ->
+    existingGutter = @_getExistingGutterWithName gutterName
+    if existingGutter
+      throw new Error 'Tried to create a gutter with a name that is already in use.'
```

Comment:
We need a convention for sorting out naming conflicts. I kinda think gutters should be prefixed with the name of the package that adds them. It's worth mentioning our policy in the docs for this.